packages/expo/cli/start/server/AsyncNgrok.ts

```diff
@@ -0,0 +1,205 @@
+  private getActivePid(): number | null {
+    return this.resolver.get()?.getActiveProcess?.()?.pid ?? null;
+  }
+
+  public getActiveUrl(): string | null {
+    return this.serverUrl;
+  }
+
+  /** Terminate the instance by the pid. */
+  private killInstance() {
+    const pid = this.getActivePid();
+    if (pid !== null) {
+      try {
+        process.kill(pid);
+      } catch (e) {
+        Log.error(`Failed to kill ngrok tunnel PID: ${pid}`);
+      }
+    }
+  }
+
+  /** Exposed for testing. */
+  async _getProjectHostnameAsync() {
+    // TODO: Maybe assert no robot users?
+    const username = getActorDisplayName(await getUserAsync());
+
+    return [
+      // TODO: Is this needed?
+      await this.getProjectRandomnessAsync(),
+      slugify(username),
+      // Use the port to distinguish between multiple tunnels (webpack, metro).
+      this.port,
+      NGROK_CONFIG.domain,
+    ].join('.');
+  }
+
+  async startAsync({ timeout }: { timeout?: number } = {}): Promise<void> {
+    // Ensure that force quitting can clean up ngrok.
+    installExitHooks(() => {
+      this.killInstance();
+    });
+
+    // Ensure the instance is loaded first, this can linger so we should run it before the timeout.
+    await this.resolver.resolveAsync({
+      // For now, prefer global install since the package has native code (harder to install) and doesn't change very often.
+      prefersGlobalInstall: true,
+    });
+
+    // Ensure ADB reverse is running.
+    if (!(await startAdbReverseAsync([this.port]))) {
+      // TODO: Better error message.
+      throw new CommandError(
+        'NGROK_ADB',
+        `Cannot start tunnel URL because \`adb reverse\` failed for the connected Android(s).`
+      );
+    }
+
+    this.serverUrl = await this._connectToNgrokAsync({ timeout });
+
+    Log.debug('[ngrok] Tunnel URL:', this.serverUrl);
+    Log.log('Tunnel ready.');
+  }
+
+  public async stopAsync(): Promise<void> {
+    Log.debug('[ngrok] Stopping Tunnel');
+
+    await this.resolver.get()?.kill?.();
+    this.serverUrl = null;
+  }
+
+  /** Exposed for testing. */
+  async _connectToNgrokAsync(
+    options: { timeout?: number } = {},
+    attempts: number = 0
+  ): Promise<string> {
+    // Attempt to stop any hanging processes, this increases the chances of a successful connection.
+    await this.stopAsync();
+
+    // Get the instance quietly or assert otherwise.
+    const instance = await this.resolver.resolveAsync({
+      shouldPrompt: false,
+      autoInstall: false,
+    });
+
+    let timer: NodeJS.Timeout | null = null;
+    const results = await Promise.race([
+      // Returns false to try again.
+      this.connectToNgrokInternalAsync(instance, attempts),
+
+      // A timeout that correctly surfaces the error to the correct lexical scope.
+      new Promise<false>((_, reject) => {
+        timer = setTimeout(() => {
+          reject(new CommandError('NGROK_TIMEOUT', 'Ngrok tunnel took too long to connect.'));
```

Comment:
ngrok is written lowercased in its own docs:
```suggestion
          reject(new CommandError('NGROK_TIMEOUT', 'ngrok tunnel took too long to connect.'));
```